P021 state based extension

flashmark opened the pull request:
As discussed previously I reworked P021 to support my floor heating pump functionality. This should be a backwards compatible upgrade from previous versions. I tried to make it as flexible as possible by adding several configuration options. Documentation is updated accordingly.

flashmark committed:
Added state extension

flashmark committed:
Fix backwards compatibility issue and tidy up

TD-er commented on src/_P021_Level.ino:
```diff
@@ -18,24 +22,119 @@
+// Make time interpretation easier by converting millis to the longer periods we want to check
+#define millis2seconds(x) ((x) / ((ulong)1000))
+#define millis2minutes(x) ((x) / ((ulong)1000 * 60))
+#define millis2hours(x) ((x) / ((ulong)1000 * 60 * 24))
```
First question: Why???

Second question:
Please take another look at the code ;)

TD-er replied:
Also these are integer operations and thus it will always round down.

For example: 59999 msec will be rounded to 0 when calling the `millis2minutes` macro.
Is that what you intend to happen?

flashmark replied:
At several places the configuration parameters are set in seconds/minutes/hours. Entered time values are used as a minimum value for an interval. Or phrased different compare if the interval is above the rounded up value. I don't like repeating the same conversion so I made them macros. In hind sight it is more reliable to compare in the millisecond domain. So defining minutes2millis and use it at the other side of the compare..

flashmark replied:
Reworked way of handling time values. Now delays are stored in PCONFIG_ULONG as seconds. For user interface convert these for convenient human interpretation (two scales will be provided). For the algorithm the seconds are compared to millis(). The time scaling is done by reworked macros.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -60,70 +159,171 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+      P021_FLAGS            = 0;                   // Reset all flags
+      UserVar.setFloat(event->TaskIndex, P021_VALUE_STATE, (float)P021_STATE_IDLE);
+      P021_remote[event->TaskIndex]    = 0;        // Remote control state is "off"
+      P021_timestamp[event->TaskIndex] = millis(); // Start counting time from now on
```
Function `PLUGIN_SET_DEFAULTS` is called only once, when initially adding the plugin in the task, so storing the millis() at that time is not useful

flashmark replied:
Correct, just an attempt of "defensive programming" initializing all static/persistent data. I will save some memory space by removing it.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -60,70 +159,171 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+        P021_control_state control_state = (P021_control_state)UserVar.getFloat(event->TaskIndex, P021_VALUE_STATE);
+
+        // Add some debug information
+        String msg = F("State= ");
+        msg += P021_printControlState(control_state);
+        msg += (F(", Pump= "));
+        msg += (int)UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT) == 1 ? F("on") : F("off");
+        msg += F(", Remote= ");
+        msg += P021_remote[event->TaskIndex];
+        msg += F(", Timer= ");
+        msg += (int)(millis2seconds(timePassedSince(P021_timestamp[event->TaskIndex])));
+        addFormNote(msg);
```
Please use `strformat()` function to build complex (log) strings.

TD-er replied:
Also you should not use very use case specific terms in this plugin.
The level plugin should not care/know whether it is driving a pump, some motor, a (water) valve, or anything else.
So better change "Pump" into "Output".

flashmark replied:
Thanks for spotting some left-overs.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+  #ifdef PLUGIN_021_DEBUG
+
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
+    String log = F("P021: Calculated State= ");
+    log += P021_printControlState(new_control_state);
+    log += F(" -> ");
+    log += F("; relay= ");
+    log += relay_output;
+    log += F("; timer= ");
+    log += timePassedSince(timestamp) / 1000;
+    log += F("; mode= ");
+    log += P021_printControlMode(P021_OPMODE);
+    log += F("; value= ");
+    log += value;
+    log += F("; remote= ");
+    log += remote_state;
+    addLogMove(LOG_LEVEL_DEBUG, log);
```
Use `strformat()` please.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Retrieve one bit flag from a set of flags stored in one UINT16
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+boolean P021_getConfigBit(uint16_t config, int pos)
+{
+  boolean res = ((config >> pos) & 0x0001) == 1;
+
+  return res;
+}
```
Arduino provides a `bitRead()` function/macro, no need to re-invent that wheel.

flashmark replied:
Thanks, replaced.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Set one bit flag in a set of flags sored in one UINT16
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+uint16_t P021_setConfigBit(uint16_t config, int pos, boolean val)
+{
+  uint16_t mask = (0x0001 << pos);
+
+  if (val)
+  {
+    config |= mask;
+  }
+  else
+  {
+    mask   ^= 0xFFFF;
+    config &= mask;
+  }
+  return config;
```
Ditto for `bitWrite()`.

flashmark replied:
Makes code a lot more readable :-)

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+    digitalWrite(P021_GPIO_RELAY, relay_output ? HIGH : LOW);
+  }
+
+  #ifdef PLUGIN_021_DEBUG
+
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
```
`PLUGIN_021_DEBUG` can only be enabled if `BUILD_NO_DEBUG` is _not_ defined, causing the build to fail...

flashmark replied:
I hope I did it right. Please review carefully.

TD-er commented:
See this screenshot from your docs:
![image](https://github.com/user-attachments/assets/b359f7a3-fc6a-45c3-87e1-716ebb1b8cf1)

These units of time might be perfect for your specific use case, but keep in mind the "level" plugin is a generic plugin.
So there might be enough use cases where a time set in hours (or even minutes) is too long.
Also it is quite confusing to have different units of time on the same page.

Better have it set as seconds.
Maye later we could add some input filter which can then be used on each input field where you set a time in seconds, which allows some input formatting like "1d3h5m10s" or whatever notation you wish.
The `PLUGIN_WEBFORM_SAVE` can then parse this format and convert it into seconds when saving.

But that's for later, so we can make it generally available for more plugins.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -60,70 +159,171 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+      if (P021_getConfigBit(P021_FLAGS, P021_EXT_FUNCT))
+      {
+        // FormSelector with all operation mode options
+        String options[P021_OPMODE_SIZE]      = { F("Classic"), F("Off"), F("Standby"), F("On"), F("Local"), F("Remote") };
+        int    optionValues[P021_OPMODE_SIZE] =
+        { P021_OPMODE_CLASSIC, P021_OPMODE_OFF, P021_OPMODE_STANDBY, P021_OPMODE_ON, P021_OPMODE_TEMP, P021_OPMODE_REMOTE };
+        addFormSelector(F("Control mode"), F(P021_GUID_OPMODE), P021_OPMODE_SIZE, options, optionValues, P021_OPMODE);
```
- Using String for these arrays isn't very efficient, `const __FlashStringHelper * options[] = { F("Classic"), F("Off"), F("Standby"), F("On"), F("Local"), F("Remote") };` is the common way of doing that.
- No need to use a length initializer for the arrays, the compiler will take care of that. And we have a macro `NR_ELEMENTS(optionValues)` that will calculate the actual length at compile-time for use as the `addFormSelect` length argument, so no issues if you add or remove an option from the list.
- Using `const int optionValues[] =...` instead of just `int` allows the compiler to apply a few more optimizations.

flashmark replied:
@TD-er: Agreed there might be use cases for smaller delays. However it is also not user friendly to enter a few days in seconds. With a value of 4 days I have to enter 345600 seconds... As long as the unit is clearly mentioned I don't expect much confusion. I am thinking about a compromise here.

flashmark replied:
Redesigned the way how to handle timers. Now timing parameters are stored in seconds. Two modes of presentation: standard, long timers. Standard mode will show seconds. Long timers mode will show minutes and hours with scaling in WEBFORM_LOAD and WEBFORM_SAVE. Control algorithm will scale seconds to millis() domain.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+    bool  isZero             = essentiallyZero(hysteresis);
+    float valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
+    float valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
+    return invert ? !definitelyGreaterThan(value, valueLowThreshold) : !definitelyLessThan(value, valueHighThreshold);
+  }
+  else
+  {
+    return invert ? (value < setpoint) : (value > setpoint);
```
Why not also use `definitelyGreaterThan()/definitelyLessThan()` in the second return?

flashmark replied:
I did not want to touch the existing algorithm. I learned that one shall not test for equal/unequal with floats due to inaccuracy errors; but comparing them is ok. I assume the operator will define a good hysteresis for the controller anyway as signal noise is probably a bigger issue. But if I am wrong here I can add the function.

tonhuisman replied:
Well, using 2 different types of comparisons for variables of the same magnitude in a single function seems a bit weird, IMHO. The original code uses the `definitelyGreaterThan()/definitelyLessThan()` functions, so why not use them also for the added comparisons?

tonhuisman replied:
That function can probably be simplified to:
```c++
bool P021_check_on(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force)
{
  if (force)
  {
    return true;
  }

  float valueLowThreshold  = setpoint;
  float valueHighThreshold = setpoint;

  if (symetric)
  {
    const bool isZero = essentiallyZero(hysteresis);
    valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
    valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
  }
  return invert ? !definitelyGreaterThan(value, valueLowThreshold) : !definitelyLessThan(value, valueHighThreshold);
}
```
And similarly for the `P021_check_off()` function.

flashmark replied:
I was thinking about the same style of simplification :-)

flashmark replied:
I refactored it using the overview in the documentation.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+    bool  isZero             = essentiallyZero(hysteresis);
+    float valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
+    float valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
+    return invert ? !definitelyLessThan(value, valueHighThreshold) : !definitelyGreaterThan(value, valueLowThreshold);
+  }
+  else
+  {
+    return invert ? (value > (setpoint + hysteresis)) : (value < (setpoint - hysteresis));
```
Ditto

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+// Convert control state to string for debugging purposes
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef PLUGIN_021_DEBUG
+String P021_printControlMode(int mode)
```
You're returning a flash string (`F()` macro), but have it converted to a `String` because of the return type defined. Better declare this as a `const __FlashStringHelper *`

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+{
+  switch (mode)
+  {
+    case P021_OPMODE_CLASSIC: return F("Classic"); break;
+    case P021_OPMODE_OFF:     return F("Off"); break;
+    case P021_OPMODE_STANDBY: return F("Standby"); break;
+    case P021_OPMODE_ON:      return F("On"); break;
+    case P021_OPMODE_TEMP:    return F("Local"); break;
+    case P021_OPMODE_REMOTE:  return F("Remote"); break;
+    default:                  return F("***ERROR***"); break;
```
No need for a `break` after `return`.

flashmark replied:
Kind of defensive programming. Will remove it to conform to ESPeasy style.

tonhuisman replied:
Once a `return` would fail to return we've got bigger issues to worry about 😆

flashmark replied:
I learned (a very long time ago) to keep the complex coding structures as plain as possible as forgetting a break or other part of standard structures may lead to difficult to spot issues. Same teacher told to always add a default and never put returns halfway your code. The latter I do break :-) And for optimizations: "trust the compiler".  For embedded we often break all these rules. I adapted the code to ESPeasy style.

tonhuisman replied:
Well, not to have _any_ code after a `return` is just a plain C++ (and many other languages) rule, not specific to ESPEasy 😉

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+// Convert control state to string for debugging purposes
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef PLUGIN_021_DEBUG
+String P021_printControlState(int state)
+{
+  switch (state)
+  {
+    case P021_STATE_IDLE:     return F("Idle"); break;
+    case P021_STATE_HEATING:  return F("Heating"); break;
```
Ditto for both return type and unneeded `break`.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+            remote_value = 0;
+            success      = true;
+          }
+          else
+          {
+            log += F(" **no valid value**");
+          }
+          P021_remote[event->TaskIndex] = remote_value;
+        }
+        else
+        {
+          log += F(" **no valid command**");
+        }
```
Not sure if it's needed/useful to tell the user something was wrong, as returning `false` from the `PLUGIN_WRITE` function will already log ` Info   : Command unknown: <your command>`.

flashmark replied:
Thanks. I am quite inexperienced/unsure when using the various options in the event handling. Used if for debugging what my code was doing. Save some memory, loose a tiny bit of information about what part is failing.

tonhuisman replied:
You could even consider only logging the message on `if (success)`, so the standard `Command unknown ...` message is the only output, or the user might expect the command was executed, as some don't read/interpret these messages very well... (ask me how I know... 😛)

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+      // * levelcontrol,remote, [on|off]
+
+      // parse string to extract the command
+      String command = parseString(string, 1); // already converted to lowercase
+      String subcmd  = parseString(string, 2);
+      String value   = parseString(string, 3);
```
Defining these variables (and a lot of others that don't change after the initial assignment) as `const` will help the compiler optimize the code.

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -18,24 +22,119 @@
 # include "src/Globals/RulesCalculate.h"
 # include "src/WebServer/ESPEasy_WebServer.h"
 
+# define PLUGIN_021_DEBUG
```
The directive causes `LOG_LEVEL_DEBUG` messages, but that level is not defined for all build types. You also have to check for `BUILD_NO_DEBUG` and if not defined, you shouldn't enable `PLUGIN_021_DEBUG`.

flashmark commented:
Reworking the code. I saw quite some useful comments and lacking some time to fix it in a rush.

flashmark committed:
Reworked after pull request

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -187,32 +207,27 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
-        msg += F(", Timer= ");
-        msg += (int)(millis2seconds(timePassedSince(P021_timestamp[event->TaskIndex])));
+        String msg = strformat(F(", State= %s "), P021_printControlState(control_state));
+        msg += strformat(F(", Output= %s"), (int)UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT) == 1 ? F("on") : F("off"));
```
While `strformat()` works fine processing a `F("")` string into the %s placeholder on ESP32, that's not guaranteed to always work on an ESP8266, so a wrapper should be applied like `String(<condition> ? F("x") : F("y")).c_str()` (or on a single F("") string where applicable).

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -187,32 +207,27 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
-        msg += (int)(millis2seconds(timePassedSince(P021_timestamp[event->TaskIndex])));
+        String msg = strformat(F(", State= %s "), P021_printControlState(control_state));
+        msg += strformat(F(", Output= %s"), (int)UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT) == 1 ? F("on") : F("off"));
+        msg += strformat(F(", Remote= %d"), P021_remote[event->TaskIndex]);
+        msg += strformat(F(", Timer= %d sec"), (millis2seconds(timePassedSince(P021_timestamp[event->TaskIndex]))));
```
These lines can be combined into a single strformat call, as strformat supports a variable nr. of arguments:
```
addFormNote(strformat(F("State= %s, Output= %s, Remote= %d, Timer= %d sec"),
                      String(P021_printControlState(control_state)).c_str(),
                      String(UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT) == 1 ? F("on") : F("off")).c_str(),
                      P021_remote[event->TaskIndex],
                      millis2seconds(timePassedSince(P021_timestamp[event->TaskIndex]))));

```

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -244,42 +259,62 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+        {
+          ulong  min_time      = P021_MIN_TIME;      // Value to display for minimum timer
+          ulong  interval_time = P021_INTERVAL_TIME; // Value to display for max idling time
+          ulong  force_time    = P021_FORCE_TIME;    // Value for forces run
```
Current C++ compilation for Arduino uses/prefers `uint32_t` over `ulong` as it is more explicit.

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -736,26 +779,19 @@ void P021_evaluate(struct EventStruct *event)
+    log += strformat(F("; mode= %s"), P021_printControlMode(P021_OPMODE));
+    log += strformat(F("; value= %f"), value);
+    log += strformat(F("; remote= %d"), remote_state);
     addLogMove(LOG_LEVEL_DEBUG, log);
```
Ditto, a single `strformat()` could do this:
```
    addLogMove(LOG_LEVEL_DEBUG,
               strformat(F("P021: Calculated State= %s; GPIO= %d; timer= %d sec; mode= %s; value= %f; remote= %d"),
                         String(P021_printControlState(new_control_state)).c_str(),
                         relay_output,
                         millis2seconds(timePassedSince(timestamp)),
                         String(P021_printControlMode(P021_OPMODE)).c_str(),
                         value,
                         remote_state));
```

flashmark committed:
Rework after review

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
Merge branch 'P021-state_based_extension' of https://github.com/flashmark/ESPEasy into P021-state_based_extension

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -726,7 +731,7 @@ void P021_evaluate(struct EventStruct *event)
             new_control_state = P021_STATE_ACTIVE;
           }
-          else if (timePassedSince(timestamp) >= seconds2millis(P021_MIN_TIME))
+          else if (timePassedSince(timestamp) >= (long)seconds2millis(P021_MIN_TIME))
```
Maybe the cast should be part of the macro...

flashmark committed:
Cleanup, fix init issue

flashmark committed:
Update documentation

tonhuisman commented on docs/source/Plugin/P021.rst:
```diff
@@ -58,13 +62,13 @@ State control extension
-* **Idle** : The Output (heater/pump) is swiched off. There is no reason to switch it on.
-* **Heating** : The basic comparator algorithm decides the Output must be switched on. There is a demand based upon the measured Input value.
+* **Idle** : The Output is swiched to inactive (off). There is no reason to switch it active (on).
+* **Active** : The basic comparator algorithm decides the Output must be switched to active. There is a demand based upon the the measured Input value.
```
Typo introduced: `the the`

tonhuisman commented:
To fix the build of ESP8266 Collection D, in file `define_plugin_sets.h` you can change line 1580 to read:
```
    #ifdef ESP32
      #define USES_P094  // CUL Reader
    #endif
```
This is a planned change, already applied to some other PRs for the same reason.

tonhuisman commented:
To keep the source-formatting consistent, you are strongly encouraged to use the Uncrustify plugin in VSCode, as we have an `uncrustify.cfg` file in the repo.
Assuming you have that plugin installed (it's listed on the PlatformIO page in the ESPEasy documentation), _and_ you have downloaded the `uncrustify` executable for your platform, ([linked here](https://espeasy.readthedocs.io/en/latest/Participate/PlatformIO.html#uncrustify)) you have to configure the path/filename in the plugin settings:
(`<Ctrl>-<Comma>` and search for uncrustify)
![image](https://github.com/user-attachments/assets/ae219ed1-673d-4e58-8ac2-20e7a8ac31eb)

Once that's taken care of, you can use the `<Alt>-<Shift>-F` key combination (Windows, similar on Linux/Mac) or fight-click and select Format source, to first select the desired default formatter (Uncrustify 😉), and from then on format the current source-file.

flashmark commented:
As far as I know I am using Uncrustify as described. Using "Format document using Uncrustify" does some formatting for me. SO something seems to be wrong if this code does not match expectations.

flashmark committed:
Minor fixes

flashmark commented:
I did not found much time to work on this plugin lately. I have two open issues:
1) Seems uncrustfy does not produce what Ton expects. I think I followed the installation manual and I do see reformatting using Uncrustify when hitting the <alt><shift>F key. How can I check if teh right formatting rules are loaded?
2) More severe is running out of flash memory for at least collection C. I am trying to reduce the footprint. But I added quite some new functionality and associated parameters. I can cut functionality for some builds, but I wonder if that is useful. I can also move to a new plugin and make it optional.

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

tonhuisman commented:
> 1. How can I check if teh right formatting rules are loaded?

The VSCode-settings for Uncrustify point to the (default) `./uncrustify.cfg` file, and assuming that file is in the root of the source tree you use, it should be loaded automatically. You should be using the latest uncrustify.exe (assuming you're on Windows), but that's not updated recently (couple of years ago, actually), and also VSCode as to be fairly up to date (currently v1.92.0).


> 2\. More severe is running out of flash memory

The ways to solve that are reducing code size and disabling some features for builds that have `LIMIT_BUILD_SIZE` defined, but that's set for many of the ESP8266 builds...

First things to exclude from these builds is non-error logging, like in the PLUGIN_WRITE case (lines 486/492), check line 426 for an example. (That'll save into the dozens/hundreds of bytes, but not the ~1.3k that needs to be shaved off 🤔) (-/- 120 bytes)
Same for non-essential UI messages, like some `addFormNote()` lines.

And you have optional short/long time notation, that's also adding some non-required code, you could exclude the 'expensive' parts for limited builds. (-/- 312 bytes)
Shortening strings like `seconds` to `sec` etc. scrape off a few bytes here and there.

Handling command parameters "on"/"off" could be replaced by checking for 0/1 (like many other plugins do), as these numeric values will be available in `event->Par2`, so no need to parse out the `value` String, and do a relatively elaborate string-compare there, where `if (event->Par2 == 0)` could be used. You might want to check if `const bool hasValue = !parseString(string, 3).isEmpty();` is true, as providing no value also results in 0, but that could also be a feature... 😉 (-/- 48 bytes)

Avoiding nearly duplicate code like `P021_check_on()` vs. `P021_check_off()`, that also have quite some room for optimization, will result in smaller bin size.
Reworked those a bit (comments not included here, and 'single-lined' the `if force` check for space reasons only)
```c++
float P021_symetric_threshold(float setpoint, float hysteresis, bool invert) {
  const bool isZero = essentiallyZero(hysteresis);
  const float delta = (isZero ? 1.0f : (hysteresis / 2.0f));

  return setpoint + ((invert ? -1.0f : 1.0f) * delta);
}
bool P021_check_on(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force) {
  if (force) { return true; }

  const float Threshold = symetric 
    ? P021_symetric_threshold(setpoint, hysteresis, invert)
    : setpoint; // Default to Setpoint for asymetric hysteresis

  return invert ? definitelyLessThan(value, Threshold) : definitelyGreaterThan(value, Threshold);
}
bool P021_check_off(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force) {
  if (force) { return false; }

  const float Threshold = symetric
    ? P021_symetric_threshold(setpoint, hysteresis, invert)
    : setpoint + ((invert ? -1.0f : 1.0f) * hysteresis);

  return invert ? definitelyGreaterThan(value, Threshold) : definitelyLessThan(value, Threshold);
}
```
This saved me 48 bytes, compiled for ESP8266 😸

Total gain with a few more optimizations was 560 bytes for ESP8266 builds. Not enough to make the Collection C build fit, but a nice improvement already 👍

tonhuisman commented:
Here's a .diff file with my patches: 
[P021_tonhuisman.zip](https://github.com/user-attachments/files/16559414/P021_tonhuisman.zip)

Extract the .diff to the root of your branch and use command `git apply P021_tonhuisman.diff` to see what I changed. (Original P021 source should be unchanged)

flashmark committed:
P021 Optimize memory footprint

flashmark commented:
Thanks for the suggestions. I already started some optimizations and reused most of the suggestions. By removing the write functionality It just seems to fit for the targets I tried. If we need to remove more functionality it might be easier to revert to the original P021 and create a new plugin. Gives me also the opportunity to remove backwards compatibility.

Question: is it possible to define the F() strings once when they are duplicated? Some units like "seconds" and "minutes" are shown multiple times. Of has the compiler some de-duplication magic onboard?

I did not test the changes yet. First test shows that the hysteresis algorithm does not work anymore. I will try to find some time coming week to debug the issue. Please don't merge to mega yet!

tonhuisman commented:
> Question: is it possible to define the F() strings once when they are duplicated? Some units like "seconds" and "minutes" are shown multiple times. Of has the compiler some de-duplication magic onboard?

The compiler does de-duplication on those strings as much as possible, so no need to worry about that.

flashmark committed:
P021 fixed issue in control algoritm

TD-er commented on src/_P021_Level.ino:
```diff
@@ -918,8 +918,8 @@ bool P021_check_off(float value, float setpoint, float hysteresis, bool invert,
   const float Threshold = symetric
-    ? P021_symetric_threshold(setpoint, hysteresis, invert)
-    : setpoint + ((invert ? -1.0f : 1.0f) * hysteresis);
+    ? P021_symetric_threshold(setpoint, hysteresis, !invert)
```
Ah that's a tricky bug....

flashmark committed:
Merge branch 'mega' into P021-state_based_extension

flashmark committed:
P021 Squeeze out some flash usage

TD-er commented on src/_P021_Level.ino:
```diff
@@ -312,10 +312,10 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
 
           # ifndef P021_MIN_BUILD_SIZE
+          String unit1 = F("seconds");                 // use minutes or seconds
+          String unit2 = F("seconds");                 // use hours or seconds
```
I don't think this is going to save a lot in the build size, to leave it out.
You can however make the code a bit more efficient by either declaring it as type `const __FlashStringHelper *` instead of `String`.
The `addUnit()` function has a version which accepts `const __FlashStringHelper *` as type, so no conversion is needed.

And it can even be shorter as `[s]` is also quite clear what unit it represents plus there is also a version of the `addUnit` function which accepts a char.

So you can simply use `addUnit('s');` or 
```c++
const char unit1('s');
addUnit(unit1);
```

flashmark replied:
I decided to remove the option to switch between long/short time span units. This is part of the construction I needed to support both. these remarks helps learing more about the __FlashStringHelper.

tonhuisman replied:
You can have that option, but it won't fit in an ESP8266 build. so better to exclude it from there. Removing it al together is a bit drastic, but still a solution in fighting the .bin-size issue.

flashmark replied:
Indeed that is what I intended. Comment out (#ifndef) the code for the option. This removes the selector on the webform and the code to convert from/to seconds. Part of this stripping was to strip code to select the strings for the units. In the end I dropped the units completely in the build size limited version. This got me 50 bytes from the edge for collection C. When not size limited all original functionality is still available. (makes the code a bit less readable, but it fits for now). If we have to squeeze more we get close to revert to the original code and make this a new plugin. Unless we can still find real code optimizations. 

With respect to moving to a character only: A bit poor man's solution to go for single character. I would have done this in the build size limited code, but not showing it at all brought me a bit more. Seconds should be the standard SI unit for time...

tonhuisman replied:
The last 200 bytes are the hardest to get rid of. 😱 
In another PR I shaved off ~200 bytes of P038 (also in `normal` builds), that might get your changes to fit.
I might apply similar fixes to other plugins as well.

flashmark commented:
Seems I am doing a Sisyphus job in shrinking down the flash contents. For now 8266 collection C should fit again, but at the cost of some functionality in minimal build. 
Also quite interesting which tricks do help and which seems to be optimized already by the compiler. Calling millis() multiple times is cheaper than copying a const uint32_t value.

TD-er commented on src/_P021_Level.ino:
```diff
@@ -18,39 +22,179 @@
+
+// Static storage for global state info. Track per ESPeasy plugin instance
+static bool P021_remote[TASKS_MAX];        // Static storage for remote control.
+static uint32_t P021_timestamp[TASKS_MAX]; // Static storage for timestamp last change
```
Why not use a PluginStruct to store these values?
That's why I created those, so you don't need to allocate these kind of arrays which are hardly ever used but still take up precious memory.

A few very simple and small examples: 
- P024_data_struct.h/.cpp
- P060_data_struct.h/.cpp
- P069_data_struct.h/.cpp
- P107_data_struct.h/.cpp


Also these are not initialized to 0, so you may end up with strange results if you load the webform while the task was never initialized.

flashmark replied:
I know these structures. I copied the pattern from the original plugin. Seems we are lacking flash memory more than ram memory and adding a complete class probably requires more code in flash. But I will look at it. Both variables don't show on the webform. They are initialized when the plugin is initialized and used to track data between calls.

flashmark replied:
Sorry, timer initialization got lost during the optimizations.

flashmark committed:
P021 timer initialization

TD-er commented:
> Seems I am doing a Sisyphus job in shrinking down the flash contents. For now 8266 collection C should fit again, but at the cost of some functionality in minimal build. Also quite interesting which tricks do help and which seems to be optimized already by the compiler. Calling millis() multiple times is cheaper than copying a const uint32_t value.

I know exactly what you mean.
And now the most frustrating part...
When you think you found something which is saving bytes in the bin file, you tend to apply it multiple times and suddenly the build gets larger again as the compiler's optimization suddenly decides to do some loop-unrolling or something else you absolutely do not expect.

I guess you now may understand why I consider the ESP8266 to be 'feature frozen' ?

flashmark commented:
Outch, the collecion C with VCC requires more flash :-(. However I found some big fishes. Seems the compiler does not optimize larger sets of PCONFIG accesses. Indexing and offsetting requires quite some program memory. I cached several of them on places where they were accessed multiple times. Especially the flags that were read/updated using bitread/bitwrite the PCONFIG gained a lot :-) We have some bytes headroom! I need to test if code is still working as intended.

TD-er commented:
Can you pinpoint which changes you have in mind?
Maybe I can help.

flashmark commented:
I already made them, just lacking time to do the full test. I will upload them to the pull request as long as it won't be integrated until I finished testing.

flashmark committed:
P021 another set of shaving flash memory consumption

TD-er commented on src/_P021_Level.ino:
```diff
@@ -244,19 +244,20 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
         const P021_control_state control_state = (P021_control_state)UserVar.getFloat(event->TaskIndex, P021_VALUE_STATE);
 
         // Add some debug information
         const String outpstring = (int)UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT) == 1 ? F("on") : F("off");
```
C-style casts of a float to an int may not be what you'd expect.
For example 0.999999f will be a 0 when cast like this.

I assume you expect either 0 or 1, so you can also reverse the check and use the function `essentiallyZero(UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT)) ? F("off") : F("on");`
And if it is really about reducing build size, you might also do this:
```c++
const char* outpstring = essentiallyZero(UserVar.getFloat(event->TaskIndex, P021_VALUE_OUTPUT)) ? "off" : "on";
```
This way you don't need to use `.c_str()` later and since the flash strings are shorter than 4 bytes it will not increase RAM usage. ("off" is 4 bytes, including the 0-terminator, which is the same size as a pointer to a flash string)

This will not be used in the flash-string deduplication at compile time, but since the string is less than 4 bytes, it doesn't really matter.

flashmark replied:
Indeed this is reading back the output value which is assumed to be in a boolean/pin state format. This is for debugging and is skipped at minimal build. I copied part of this construction from a remark from Ton. Except I declared the String explicitly (to make to a bit more readable). The char[] is a good suggestion, but makes it even more outside the standard patterns.

flashmark replied:
Fixed the cast. Will keep the String (matter of taste)

TD-er commented on src/_P021_Level.ino:
```diff
@@ -244,19 +244,20 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
-                                      outpstring.c_str(),
-                                      P021_remote[event->TaskIndex],
-                                      millis2seconds(timePassedSince(P021_timestamp[event->TaskIndex])));
+        String msg              = strformat(F("State= %s, Output= %s, Remote= %d, Timer= %d sec"),
```
Assigning the result of `strformat` to a non-const varialble will result in both more memory usage and larger build file compared to when you call it directly where you use it. (assuming you only use it once)

```c++
addFormNote(strformat(...));
```
This will not create a copy of the string on the stack, but directly feed it to the `addFormNote` function.
After the function returns the temporary created String object is destructed and thus not taking up any memory anymore.

TD-er commented on src/_P021_Level.ino:
```diff
@@ -244,19 +244,20 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
     case PLUGIN_WEBFORM_LOAD:
     {
-      const bool extFunc = bitRead(P021_FLAGS, P021_EXT_FUNCT); // Extended functionality selected
+      const uint16_t flags   = P021_FLAGS;                     // shortcut to read existing configuration flags
```
This indeed does save quite a bit in build size, as the macro would otherwise be expanded quite a lot of times.

flashmark replied:
Yeah, I abuse one of the PCONFIG parameters to store a bunch of booleans, By the way, I made a slip of the pen at the webform_save. Forgot to set one flag using the new variable. Will upload the change soon.

TD-er commented on src/_P021_Level.ino:
```diff
@@ -639,15 +647,17 @@ const __FlashStringHelper* P021_printControlMode(int mode)
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 void P021_check_autosafe(struct EventStruct *event)
```
Maybe also fix the typo in the function name ;)

tonhuisman commented on src/_P021_Level.ino:
```diff
@@ -60,111 +204,290 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
-      P021_AUTOSAVE_TIMER      = getFormItemInt(F("pautosave")) * 60; // Store in seconds
-      P021_INVERT_OUTPUT       = isFormItemChecked(F("inv")) ? 1 : 0;
+      const bool newExtFunct  =  (getFormItemInt(F(P021_GUID_EXT_FUNCT)) != 0);
+      uint16_t   flagRegister = P021_FLAGS; // Reduce expensive access to P021 to write new flags
```
This variable can also be a `const`, and for consistency should be renamed to `flags`. The other `flags` variable is not in this scope, so no conflicts.

flashmark replied:
This is not a const as it is updated by successive bitwrite() calls. This is where most of the flash memory is saved. Indeed, forgot to rename it when reviewing it myself.

flashmark committed:
P021 fixed an oops and reworked review comments

flashmark commented:
New commit, tested on a ESP32. I have to setup something to test it on ESP8266.

TD-er commented:
@flashmark I just want to say thank you for giving me some ideas to look at ways to reduce bin size.
I am now working on P001_switch and looks like I can reduce it already by roughly 12k and haven't even started... ;)
Probably similar results can be achieved in P009 and P019.

flashmark commented:
Nice to hear I can contribute to a great project. I like to play with my little ESPeasy projects. And now writing my own stuff I see it has a pretty good architecture below the surface. Nice balance between a good structure and still thinking about the limitations of the boards. Although for now I start to get a bit nervous about failing build due to code size myself :-) So reducing P001 will give some room for P021...

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

TD-er commented:
OK, I have no idea what values I have been looking at when concluding it would save 10k+....
Ton also tested my code and I also did a lot more testing and the gain is a lot less.
However I am convinced my code changes will make the code (for P001/P009/P019) more readable and probably a bit faster too.

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

TD-er commented:
I just cancel your build job and after the release build has finished, I will resume your GH Actions build
Otherwise I have to stay up a bit longer...

TD-er committed:
Merge branch 'mega' into P021-state_based_extension

TD-er commented:
I just updated this PR to include merged I made right after the release build.
This should give a few kB extra room for ESP8266 builds.
Please do not fill those up immediately.

flashmark commented:
I will only try to reduce memory usage. With the memory reduction all builds are successful. I tested the plugin with and without reduced build size on a ESP32. I am not sure if tests on ESP8266 are required for a merge to mega. If such tests are required I need some time to arrange a test setup. Most of my spare modules are ESP32 and ESP32-C3. This weekend I will most time far away from my keyboard.

tonhuisman commented:
It's good to know if all is working as intended on an ESP8266, but if it already works fine on an ESP32, then most of the time it won't be problematic on an ESP8266. This plugin is not directly related to CPU performance or available memory, so not too much to worry about.

TD-er commented:
The most important thing to keep in mind is that you should only add new features of existing plugins to ESP32-builds and only to ESP8266 when it isn't a regular build (thus only custom builds should be able to run those new features.

As you've seen yourself, it is really hard to make it all fit again and it has taken me about 20 - 24 hours of work last week to gain these extra few kBytes of space in the build size of ESP8266.

Like I've mentioned before a number of times, I have spent many 100's of hours in the past to make it all fit on ESP8266 and I don't think I should keep doing this as it will result in hours lost which could have been spent on new features, bug fixes, etc.

Espressif already declared the ESP8266 as deprecated about 5 years ago, so we should just consider the ESP8266 as "feature frozen" for regular builds like "collection X" or "normal" etc.
New features should -where possible- still be possible on ESP8266 for "custom" builds, but no longer for other builds.

flashmark commented:
Sorry I am late in reply. Other stuff to do outside ESPeasy.

Problem with my changes to P021 was the fact I refactored it a lot to fit the new functionality. This makes it quite difficult to get the same memory footprint by "disabling" the new stuff. Is there a way to detect the ESP8266 builds to restrict the functionality even more? We could at least disable all options on the setup page and cut out the associated control part too. Would make documentation a bit more difficult as documented features are not available... 

Discussion started with an new plugin fitting my timing related requirements. Suggestion was to add them to P021. Indeed it was a nice challenge to get it combined with backwards compatibility, but at the cost of some memory. We can still decide to create a new plugin and leave P021 unaltered. It will duplicate functionality, but the new plugin can be left out of all ESP8266 code.

With additional savings by cutting functionality on ESP8266 I think it is wise to do some testing with this code on a 8266 too. For now the optimized code is tested on ESP32 only. Please advise how to continue.

tonhuisman commented:
Currently all builds succeed, because we where able to (again) reduce the ESP8266 builds by a few kB (though I think we're close to rock-bottom now). I'd leave the new functionality in P021, and not create a new plugin.
It seems to work as intended, so no need to change again, it's fine as it is now 👍.

TD-er commented:
I think you should add a separate define which will then be used only within the P021 code.

```c++
#ifndef LIMIT_BUILD_SIZE
#define FEATURE_P021_EXTRAS  1
#else 
#define FEATURE_P021_EXTRAS  0
#endif
```
And this could then be used as wrapper for all these extra features like this:
```c++
#if (FEATURE_P021_EXTRAS)
....
#endif
```

In the documentation you could add some asterisk next to the title of some feature which is only included in non-limited builds.

flashmark committed:
[P021] disable new feaures for low memory builds. Fix issues.

flashmark committed:
[p021] fixed unused variable

flashmark committed:
[P021] Documentataion update to refect limited build versions

flashmark commented:
I updated the code to remove new functionality on builds with limited flash sizes. Still to determine how to switch this on for all affected builds. Documentation is also updated to mark the unavailable parameters. Functionality is hand-tested for full & limited variants on ESP32.

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark commented:
I could not spend much time on this plugin for some time. Meanwhile the plugin is working for more than a week to control my floor heating pump without problems. Can somebody review the logic to switch the additional functionality on/off and advise on a final configuration? I am missing some experience in managing multiple build options. Probably we should left all esp8266 versions without new functionality and enable it on the other hardware.

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
[P021] Bugfix

flashmark commented:
Found a bug introduced due to the memory optimizations. I missed a condition check due to which extending the active phase after input became inactive was not working. I did manual testing on ESP8266 (Wemos D1) and ESP32. Software has been running for weeks at my place without issues except for not extending the active phase as described above. To be reviewed is the selection of the code size and supported functionality. For now it depends on LIMIT_BUILD_SIZE.

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

TD-er committed:
Merge branch 'mega' into P021-state_based_extension

TD-er committed:
Merge branch 'mega' into P021-state_based_extension

TD-er committed:
Merge branch 'mega' into P021-state_based_extension

TD-er committed:
Merge branch 'mega' into P021-state_based_extension

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
Merge branch 'letscontrolit:mega' into P021-state_based_extension

flashmark committed:
Minor style change

TD-er committed:
Merge branch 'mega' into P021-state_based_extension
